### PR TITLE
[mod] start putting timeout in certificate code

### PR DESCRIPTION
**Pull request by Psycojoker**
As discussed with @julienmalik and @alexAubin, I'm starting this PR so we can discuss about how we should handle timeouting requests in the cert management.

### CODE REVIEW

File: src/yunohost/certificate.py [hunk 1]
```diff
@@ -567,7 +569,11 @@ def _fetch_and_enable_new_certificate(domain, staging=False):
         raise MoulinetteError(errno.EINVAL, m18n.n(
             'certmanager_cert_signing_failed'))
 
-    intermediate_certificate = requests.get(INTERMEDIATE_CERTIFICATE_URL).text
+    try:
+        intermediate_certificate = requests.get(INTERMEDIATE_CERTIFICATE_URL, timeout=30).text
+    except Timeout:
+        # XXX what should we do here? retry?
```
Comment (alexAubin) [inline src/yunohost/certificate.py]: If we didn't get the intermediate certificate, the certificate won't work, so I'd say we just do a `raise MoulinetteError(errno.EINVAL, m18n.n('certmanager_couldnt_fetch_intermediate_cert'))`

(Btw, I never know if we do care about which errno to use ?)

File: src/yunohost/certificate.py [hunk 2]
```diff
@@ -837,7 +843,10 @@ def _dns_ip_match_public_ip(public_ip, domain):
 
 def _domain_is_accessible_through_HTTP(ip, domain):
     try:
-        requests.head("http://" + ip, headers={"Host": domain})
+        requests.head("http://" + ip, headers={"Host": domain}, timeout=30)
+    except Timeout as e:
+        # XXX what should we do here? retry?
```
Comment (alexAubin) [inline src/yunohost/certificate.py]: If we really have a 30 sec timeout (which to me is crazy for what we are trying to do here), then we should return False.

But maybe we could have something like : 

- have a 10-15 second timeout first
- if it doesn't work, have something like "Warning : having trouble contacting domain. Retrying..."
- retry with a 30 second timeout
- return False if still timeouting

But I'm having a hard time imagining a situation where the domain was up and everything running properly, but it takes more than 30 seconds for a HEAD request to complete :/...

Comment (julienmalik) [inline src/yunohost/certificate.py]: 30 sec also seems a lot to me. 5 to 10 should be more than enough. Here we are trying to contact ourself, and we just want to detect the case where it does not work at all (due to hairpinning or other weird firewall setups).

File: src/yunohost/certificate.py (hunk 1 above)
Comment (julienmalik) [inline src/yunohost/certificate.py]: trusting @alexAubin  on this

File: src/yunohost/certificate.py (hunk 2 above)
Comment (julienmalik) [inline src/yunohost/certificate.py]: I may not see the big picture on this, but I would go for something way more simple : if timeout after 5 to 10 seconds, then just abort with a good error message, like "timeout reached when your server tried to contact itself via its public IP address. you may very well be experiencing hairpinning, or the firewall/router ahead of your server is misconfigured"

Comment (alexAubin) [inline src/yunohost/certificate.py]: @julienmalik Well yep, but @Psycojoker was pushing to have a 30 sec timeout 😛 . But I'd be in favor of this too.

**Comment by alexAubin** [comment]
Haven't tested the real timeout, but I tested the lines inside the exceptions 😕

**Comment by Psycojoker** [comment]
We need one more approval for this one.

File: src/yunohost/certificate.py [hunk 3]
```diff
@@ -567,7 +569,10 @@ def _fetch_and_enable_new_certificate(domain, staging=False):
         raise MoulinetteError(errno.EINVAL, m18n.n(
             'certmanager_cert_signing_failed'))
 
-    intermediate_certificate = requests.get(INTERMEDIATE_CERTIFICATE_URL).text
+    try:
+        intermediate_certificate = requests.get(INTERMEDIATE_CERTIFICATE_URL, timeout=30).text
+    except Timeout:
```
Comment (opi) [inline src/yunohost/certificate.py]: you should add `as e` to be conpliant with yunohost coding standards (tm)

File: src/yunohost/certificate.py [hunk 4]
```diff
@@ -837,7 +842,10 @@ def _dns_ip_match_public_ip(public_ip, domain):
 
 def _domain_is_accessible_through_HTTP(ip, domain):
     try:
-        requests.head("http://" + ip, headers={"Host": domain})
+        requests.head("http://" + ip, headers={"Host": domain}, timeout=10)
+    except Timeout as e:
+        logger.warning(m18n.n('certmanager_http_check_timeout'))
```
Comment (opi) [inline src/yunohost/certificate.py]: error message should include domain and ip information to be more relevant

**Comment by alexAubin** [comment]
Thanks for the feedback !

**Pull request merged**